File: src/main/java/main/http/ProxyMessage.java
package main.http;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

/**
 *
 * @author Bunnyspa
 */
public class ProxyMessage {

    public String reqMethod, reqUrl, reqVersion;
    @NotNull
    private final Map<String, String> reqHeader;
    @NotNull
    public final List<Byte> reqBody;

    public int resCode;
    public String resMsg;
    @NotNull
    private final Map<String, String> resHeader;
    @NotNull
    public final List<Byte> resBody;

    public ProxyMessage() {
        reqHeader = new HashMap<>();
        resHeader = new HashMap<>();
        reqBody = new ArrayList<>();
        resBody = new ArrayList<>();
    }

    public ProxyMessage(
            String reqMethod, String reqUrl, String reqVersion,
            @NotNull Map<String, String> reqHeader,
            int resCode, String resMsg,
            @NotNull Map<String, String> resHeader
    ) {
        this.reqMethod = reqMethod;
        this.reqUrl = reqUrl;
        this.reqVersion = reqVersion;
        this.reqHeader = new HashMap<>(reqHeader);
        this.reqBody = new ArrayList<>();
        this.resCode = resCode;
        this.resMsg = resMsg;
        this.resHeader = new HashMap<>(resHeader);
        this.resBody = new ArrayList<>();
    }

    @NotNull
    public Set<String> getReqHeaders() {
        return getHeaders(reqHeader);
    }

    @NotNull
    public Set<String> getResHeaders() {
        return getHeaders(resHeader);
    }

    public void addReqHeader(String header, String field) {
        addHeader(header, field, reqHeader);
    }

    public void addResHeader(String header, String field) {
        addHeader(header, field, resHeader);
    }

    public boolean containsReqHeader(@NotNull String header) {
        return containsHeader(header, reqHeader);
    }

    public boolean containsResHeader(@NotNull String header) {
        return containsHeader(header, resHeader);
    }

    @Nullable
    public String getReqHeader(String header) {
        return getHeader(header, reqHeader);
    }

    @Nullable
    public String getResHeader(String header) {
        return getHeader(header, resHeader);
    }

    @NotNull
    private static Set<String> getHeaders(@NotNull Map<String, String> headerMap) {
        return headerMap.keySet();
    }

    private static void addHeader(String header, String field, @NotNull Map<String, String> headerMap) {
        headerMap.put(header, field);
    }

    private static boolean containsHeader(@NotNull String header, @NotNull Map<String, String> headerMap) {
        return headerMap.keySet().parallelStream().map(String::toLowerCase).anyMatch((key) -> key.equals(header.toLowerCase()));
    }

    @Nullable
    private static String getHeader(String header, @NotNull Map<String, String> headerMap) {
        for (String key : headerMap.keySet()) {
            if (key.equalsIgnoreCase(header)) {
                return headerMap.get(key);
            }
        }
        return null;
    }

    private void fixForConnect() {
        if ("CONNECT".equals(reqMethod) && resCode == 0) {
            resCode = 200;
            resMsg = "Connection Established";
        }
    }

    @NotNull
    private static String getBodyHex(@NotNull List<Byte> byteList) {
        StringBuilder sb = new StringBuilder();
        byteList.parallelStream().forEach((b) -> sb.append(byteToHex(b)));
        return sb.toString();
    }

    @NotNull
    public String getRequestHeader() {
        String out = reqMethod + " " + reqUrl + " " + reqVersion + System.lineSeparator();
        List<String> headers = new ArrayList<>();
        reqHeader.keySet().parallelStream().map((key) -> key + ": " + reqHeader.get(key)).parallel().forEach(headers::add);
        Collections.sort(headers);
        out += String.join(System.lineSeparator(), headers);
        return out;
    }

    @NotNull
    public String getRequest() {
        String out = getRequestHeader();
        out += System.lineSeparator() + System.lineSeparator();
        if (reqBody.isEmpty()) {
            out += "(No Body)";
        } else {
            out += "Body Length: " + reqBody.size() + System.lineSeparator();
            out += getBodyHex(reqBody);
        }
        return out;
    }

    @NotNull
    public String getRepsonse() {
        String out = getRepsonseHeader();
        out += System.lineSeparator() + System.lineSeparator();
        if (resBody.isEmpty()) {
            out += "(No Body)";
        } else {
            out += "Body Length: " + resBody.size() + System.lineSeparator();
            out += getBodyHex(resBody);
        }
        return out;
    }

    @NotNull
    public static String byteToHex(byte b) {
        int v = b & 0xFF;
        return Integer.toHexString(v);
    }

    @NotNull
    public String getRepsonseHeader() {
        fixForConnect();
        String out = reqVersion + " " + resCode + " " + resMsg + System.lineSeparator();
        List<String> headers = new ArrayList<>();
        resHeader.keySet().parallelStream().map((key) -> key + ": " + resHeader.get(key)).parallel().forEach(headers::add);
        Collections.sort(headers);
        out += String.join(System.lineSeparator(), headers);
        return out;
    }

    @NotNull
    private static String getBodyString(@NotNull List<Byte> byteList) {
        byte[] byteArray = new byte[byteList.size()];
        for (int i = 0; i < byteList.size(); i++) {
            byteArray[i] = byteList.get(i);
        }
        return new String(byteArray);
    }

    @NotNull
    public String toData() {
        String out = getRequest() + System.lineSeparator();
        out += "-----" + System.lineSeparator();
        out += getRepsonse();
        return out.trim();
    }
}

File: src/main/java/main/http/ProxyMessage.kt
package main.http

import java.util.*
import kotlin.experimental.and

/**
 *
 * @author Bunnyspa
 */
class ProxyMessage {
    var reqMethod: String? = null
    var reqUrl: String? = null
    var reqVersion: String? = null
    private val reqHeader: MutableMap<String?, String>
    val reqBody: MutableList<Byte>
    var resCode = 0
    var resMsg: String? = null
    private val resHeader: MutableMap<String?, String>
    val resBody: MutableList<Byte>

    constructor() {
        reqHeader = HashMap()
        resHeader = HashMap()
        reqBody = mutableListOf()
        resBody = mutableListOf()
    }

    constructor(
        reqMethod: String?, reqUrl: String?, reqVersion: String?,
        reqHeader: Map<String?, String>,
        resCode: Int, resMsg: String?,
        resHeader: Map<String?, String>
    ) {
        this.reqMethod = reqMethod
        this.reqUrl = reqUrl
        this.reqVersion = reqVersion
        this.reqHeader = HashMap(reqHeader)
        reqBody = mutableListOf()
        this.resCode = resCode
        this.resMsg = resMsg
        this.resHeader = HashMap(resHeader)
        resBody = mutableListOf()
    }

    val reqHeaders: Set<String?>
        get() = getHeaders(reqHeader)

    fun addReqHeader(header: String, field: String) {
        addHeader(header, field, reqHeader)
    }

    fun addResHeader(header: String, field: String) {
        addHeader(header, field, resHeader)
    }

    fun containsReqHeader(header: String): Boolean {
        return containsHeader(header, reqHeader)
    }

    fun containsResHeader(header: String): Boolean {
        return containsHeader(header, resHeader)
    }

    fun getReqHeader(header: String?): String? {
        return getHeader(header, reqHeader)
    }

    fun getResHeader(header: String?): String? {
        return getHeader(header, resHeader)
    }

    private fun fixForConnect() {
        if ("CONNECT" == reqMethod && resCode == 0) {
            resCode = 200
            resMsg = "Connection Established"
        }
    }

    val requestHeader: String
        get() {
            var out = reqMethod + " " + reqUrl + " " + reqVersion + System.lineSeparator()
            val headers: MutableList<String> = mutableListOf()
            reqHeader.keys.stream().map { key: String? -> key + ": " + reqHeader[key] }
                .forEach { e: String -> headers.add(e) }
            headers.sort()
            out += java.lang.String.join(System.lineSeparator(), headers)
            return out
        }
    private val request: String
        get() {
            var out = requestHeader
            out += System.lineSeparator() + System.lineSeparator()
            if (reqBody.isEmpty()) {
                out += "(No Body)"
            } else {
                out += "Body Length: " + reqBody.size + System.lineSeparator()
                out += getBodyHex(reqBody)
            }
            return out
        }
    private val repsonse: String
        get() {
            var out = repsonseHeader
            out += System.lineSeparator() + System.lineSeparator()
            if (resBody.isEmpty()) {
                out += "(No Body)"
            } else {
                out += "Body Length: " + resBody.size + System.lineSeparator()
                out += getBodyHex(resBody)
            }
            return out
        }
    val repsonseHeader: String
        get() {
            fixForConnect()
            var out = reqVersion + " " + resCode + " " + resMsg + System.lineSeparator()
            val headers: MutableList<String> = mutableListOf()
            resHeader.keys.stream().map { key: String? -> key + ": " + resHeader[key] }
                .forEach { e: String -> headers.add(e) }
            headers.sort()
            out += java.lang.String.join(System.lineSeparator(), headers)
            return out
        }

    fun toData(): String {
        var out = request + System.lineSeparator()
        out += "-----" + System.lineSeparator()
        out += repsonse
        return out.trim { it <= ' ' }
    }

    companion object {
        private fun getHeaders(headerMap: Map<String?, String>): Set<String?> {
            return headerMap.keys
        }

        private fun addHeader(header: String, field: String, headerMap: MutableMap<String?, String>) {
            headerMap[header] = field
        }

        private fun containsHeader(header: String, headerMap: Map<String?, String>): Boolean {
            return headerMap.keys.stream().map { obj: String? -> obj!!.toLowerCase() }
                .anyMatch { key: String -> key == header.toLowerCase() }
        }

        private fun getHeader(header: String?, headerMap: Map<String?, String>): String? {
            for (key in headerMap.keys) {
                if (key.equals(header, ignoreCase = true)) {
                    return headerMap[key]
                }
            }
            return null
        }

        private fun getBodyHex(byteList: List<Byte>): String {
            val sb = StringBuilder()
            byteList.forEach { b ->
                sb.append(byteToHex(b))
            }
            return sb.toString()
        }

        private fun byteToHex(b: Byte): String {
            return Integer.toHexString((b and 0xFF.toByte()).toInt())
        }

    }
}